firedrake/mg/embedded.py

```diff
@@ -265,37 +274,44 @@ def restrict(self, gf, gc):
         """
         Vc = gc.function_space()
         Vf = gf.function_space()
-
         source_element = Vf.ufl_element()
         target_element = Vc.ufl_element()
+
+        cache = self.cache(source_element)
+        key = (Op.RESTRICT, gf, gc)
```

Comment:
Your caches will now be holding onto "big" data structures. Is memory consumption going to be a concern?

Reply:
Yes. Use: `key = (op, weakref.ref(source), weakref.ref(target))`

Reply:
Alright, I'll use `key = (op, weakref.ref(source.dat), weakref.ref(target.dat))`

Reply:
Needs manual cleaning out of the dictionary occasionally if any of the weakrefs are dead.

Reply:
Would a [`WeakKeyDictionary`](https://docs.python.org/3/library/weakref.html#weakref.WeakKeyDictionary) not be the right thing here?

Reply:
I don't think so. The whole key is a tuple, we don't want a weakref to that.

Reply:
No, for two reasons:
1. you can't take a weakref of a tuple
2. even if you could, the key is constructed in this function, so the strong reference to the key doesn't live past the end of the function, and therefore the key-value mapping would immediately be removed.